src/downloads_counter.rs

```diff
@@ -0,0 +1,149 @@
+use crate::App;
+use anyhow::Error;
+use dashmap::{DashMap, SharedValue};
+use diesel::{pg::upsert::excluded, prelude::*};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// crates.io receives a lot of download requests, and we can't execute a write query to the
+/// database during each connection for performance reasons. To reduce the write load, this struct
+/// collects the pending updates from the current process and writes in batch.
+///
+/// To avoid locking the whole data structure behind a RwLock, which could potentially delay
+/// requests, this uses the dashmap crate. A DashMap has the same public API as an HashMap, but
+/// stores the items into `num_cpus()*4` individually locked shards. This approach reduces the
+/// likelyhood of a request encountering a locked shard.
+///
+/// Persisting the download counts in the database also takes advantage of the inner sharding of
+/// DashMaps: to avoid locking all the download requests at the same time each iteration only
+/// persists a single shard at the time.
+///
+/// The disadvantage of this approach is that download counts are stored in memory until they're
+/// persisted, so it's possible to lose some of them if the process exits ungracefully. While
+/// that's far from ideal, the advantage of batching database updates far outweights potentially
+/// losing some download counts.
+#[derive(Debug)]
+pub struct DownloadsCounter {
+    inner: DashMap<i32, AtomicUsize>,
```

Comment:
Is the `AtomicUsize` here required? Maybe I'm misunderstanding the `DashMap`. It sounded to me like the `DashMap` was already handling all of the thread-sync problems for us.

Reply:
Yes, it's required.

`DashMap` does not solve all the concurrency problems we could have. What it does is reduce the amount of locked items we have when acquiring a write lock by turning `RwLock<HashMap<K, V>>` (what we'd have without the crate) into a `Vec<RwLock<HashMap<K, V>>>`, with each vector item being a shard.

If we didn't use an atomic here we'd be forced to acquire a write lock over the shard whenever we needed to update a download count. Thanks to the atomic we can instead acquire a read lock to update a download count, having to acquire a write lock only when inserting a missing version into the map.

Reply:
I see, thanks :)